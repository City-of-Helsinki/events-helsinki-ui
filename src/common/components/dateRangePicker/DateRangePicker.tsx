import 'react-datepicker/dist/react-datepicker.css';
import './datePicker.scss';

import isEqual from 'date-fns/isEqual';
import fi from 'date-fns/locale/fi';
import sv from 'date-fns/locale/sv';
import max from 'lodash/max';
import React, { FunctionComponent } from 'react';
import DatePicker, { registerLocale } from 'react-datepicker';
import { useTranslation } from 'react-i18next';

import { DATE_PICKER_INPUT_STATE } from '../../../constants';
import useLocale from '../../../hooks/useLocale';
import { formatDate } from '../../../util/dateUtils';
import DateRangeInputs from './DateRangeInputs';

registerLocale('fi', fi);
registerLocale('sv', sv);

export interface DateRangePickerProps {
  endDate: Date | null;
  isMenuOpen: boolean;
  name: string;
  onChangeEndDate: (date: Date | null) => void;
  onChangeStartDate: (date: Date | null) => void;
  startDate: Date | null;
}

const DateRangePicker: FunctionComponent<DateRangePickerProps> = ({
  endDate,
  isMenuOpen,
  name,
  onChangeEndDate,
  onChangeStartDate,
  startDate,
}) => {
  // References to input fields to set focus
  const { t } = useTranslation();
  const datePicker = React.useRef(null);
  const endDateRef = React.useRef<HTMLInputElement>(null);
  const startDateRef = React.useRef<HTMLInputElement>(null);
  // Variable to keep track on date field
  const [selectedDatePickerInput, setDatePickerInput] = React.useState<
    DATE_PICKER_INPUT_STATE
  >(DATE_PICKER_INPUT_STATE.NOT_SELECTED);
  const locale = useLocale();
  // Raw dates to be saved for input fields
  const [startDateRaw, setStartDateRaw] = React.useState<string>(
    formatDate(startDate)
  );
  const [endDateRaw, setEndDateRaw] = React.useState<string>(
    formatDate(endDate)
  );
  const [shouldChangeFocusToEnd, setShouldChangeFocusToEnd] = React.useState(
    false
  );

  const handleInputBlur = (
    selectedInput: DATE_PICKER_INPUT_STATE,
    date: Date | null
  ) => {
    switch (selectedInput) {
      case DATE_PICKER_INPUT_STATE.END_TIME_SELECTED:
        onChangeEndDate(date);
        setEndDateRaw(formatDate(date));
        break;
      case DATE_PICKER_INPUT_STATE.START_TIME_SELECTED:
        onChangeStartDate(date);
        setStartDateRaw(formatDate(date));
        break;
    }
  };

  const handleDocumentFocusin = React.useCallback(() => {
    const active = document.activeElement;
    const endDateInput = endDateRef.current;
    const startDateInput = startDateRef.current;
    // setSelected method is not added to ReactDatePicker types, but method still exists
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const current: any = datePicker.current;

    if (
      endDateInput &&
      active instanceof Node &&
      endDateInput.contains(active)
    ) {
      current?.setSelected(endDate);
    }

    if (
      startDateInput &&
      active instanceof Node &&
      startDateInput.contains(active)
    ) {
      current?.setSelected(startDate);
    }
  }, [endDate, startDate]);

  React.useEffect(() => {
    document.addEventListener('focusin', handleDocumentFocusin);
    // Clean up event listener to prevent memory leaks
    return () => {
      document.removeEventListener('focusin', handleDocumentFocusin);
    };
  }, [handleDocumentFocusin]);

  React.useEffect(() => {
    if (isMenuOpen) {
      startDateRef.current?.focus();
    }
  }, [isMenuOpen]);

  // Change focus to end date input on this hook so
  // all other states has been changed as well
  React.useEffect(() => {
    if (shouldChangeFocusToEnd) {
      endDateRef.current?.focus();
      setShouldChangeFocusToEnd(false);
    }
  }, [shouldChangeFocusToEnd]);

  const getSelectedDate = () => {
    switch (selectedDatePickerInput) {
      case DATE_PICKER_INPUT_STATE.END_TIME_SELECTED:
        return startDate;
      case DATE_PICKER_INPUT_STATE.START_TIME_SELECTED:
        return endDate;
      case DATE_PICKER_INPUT_STATE.NOT_SELECTED:
        return null;
    }
  };
  const selectedDate = getSelectedDate();
  return (
    <>
      <DatePicker
        ref={datePicker}
        locale={locale}
        minDate={max([new Date(), startDate])}
        selectsStart={true}
        selectsEnd={true}
        selected={selectedDate}
        chooseDayAriaLabelPrefix={t(
          'commons.dateSelector.accessibility.chooseDayPrefix'
        )}
        previousMonthButtonLabel={t(
          'commons.dateSelector.accessibility.previousMonth'
        )}
        nextMonthButtonLabel={t('commons.dateSelector.accessibility.nextMonth')}
        // Inline datepicker doesn't have field so keep datepicker alway open instead
        open
        popperClassName={'react-datepicker__no-popper'}
        showPopperArrow={false}
        onChange={(date: Date) => {
          switch (selectedDatePickerInput) {
            case DATE_PICKER_INPUT_STATE.START_TIME_SELECTED:
              onChangeStartDate(date);
              setStartDateRaw(formatDate(date));

              if (!startDate || !isEqual(startDate, date)) {
                setShouldChangeFocusToEnd(true);
              }
              break;
            case DATE_PICKER_INPUT_STATE.END_TIME_SELECTED:
              onChangeEndDate(date);
              setEndDateRaw(formatDate(date));
              break;
          }
        }}
        customInput={
          <DateRangeInputs
            endDate={endDate}
            endDateRaw={endDateRaw}
            endDateRef={endDateRef}
            inputName={name}
            onBlurInput={handleInputBlur}
            setDatePickerInput={setDatePickerInput}
            setEndDateRaw={setEndDateRaw}
            setStartDateRaw={setStartDateRaw}
            startDate={startDate}
            startDateRaw={startDateRaw}
            startDateRef={startDateRef}
          />
        }
        startDate={startDate}
        endDate={endDate}
      />
    </>
  );
};

export default DateRangePicker;
