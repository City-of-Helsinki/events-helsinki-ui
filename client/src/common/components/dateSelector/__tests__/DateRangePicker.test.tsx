import * as React from "react";
import renderer from "react-test-renderer";

import DateRangePicker from "../DateRangePicker";

test("DateRangePicker matches snapshot", () => {
  const component = renderer.create(
    <DateRangePicker
      endDate={new Date("2019-12-20")}
      locale="fi"
      onChangeEndDate={date => {}}
      onChangeStartDate={date => {}}
      startDate={new Date("2019-11-20")}
    />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

export {};
