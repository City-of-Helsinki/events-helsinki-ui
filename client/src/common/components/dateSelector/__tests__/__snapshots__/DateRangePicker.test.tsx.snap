// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateRangePicker matches snapshot 1`] = `
Array [
  <div
    className="react-datepicker-wrapper"
  >
    <div
      className="react-datepicker__input-container"
    >
      <div
        className="dateRangeInputsContainer"
      >
        <div
          className="dateInputWrapper"
        >
          <label>
            Alkamispäivä
          </label>
          <div
            className="formatInfo"
          >
            kk.mm.vvvv
          </div>
          <div
            className="inputWrapper"
          >
            <div
              className="input"
            >
              <input
                className=""
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                readOnly={false}
                value="20.11.2019"
              />
            </div>
            <div
              className="icon"
            >
              <svg>
                calendar-add.svg
              </svg>
            </div>
          </div>
        </div>
        <div
          className="dateSeparator"
        >
          —
        </div>
        <div
          className="dateInputWrapper"
        >
          <label>
            Loppumispäivä
          </label>
          <div
            className="formatInfo"
          >
            kk.mm.vvvv
          </div>
          <div
            className="inputWrapper"
          >
            <div
              className="input"
            >
              <input
                className=""
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                readOnly={false}
                value="20.12.2019"
              />
            </div>
            <div
              className="icon"
            >
              <svg>
                calendar-add.svg
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="react-datepicker-popper react-datepicker__no-popper"
    style={
      Object {
        "left": 0,
        "opacity": 0,
        "pointerEvents": "none",
        "position": "absolute",
        "top": 0,
      }
    }
  >
    <div>
      <div
        className="react-datepicker"
      >
        <div
          className="react-datepicker__triangle"
        />
        <button
          className="react-datepicker__navigation react-datepicker__navigation--previous"
          onClick={[Function]}
          type="button"
        >
          Previous Month
        </button>
        <button
          className="react-datepicker__navigation react-datepicker__navigation--next"
          onClick={[Function]}
          type="button"
        >
          Next Month
        </button>
        <div
          className="react-datepicker__month-container"
        >
          <div
            className="react-datepicker__header"
          >
            <div
              className="react-datepicker__current-month"
            >
              joulukuu 2019
            </div>
            <div
              className="react-datepicker__header__dropdown react-datepicker__header__dropdown--scroll"
              onFocus={[Function]}
            />
            <div
              className="react-datepicker__day-names"
            >
              <div
                className="react-datepicker__day-name"
              >
                ma
              </div>
              <div
                className="react-datepicker__day-name"
              >
                ti
              </div>
              <div
                className="react-datepicker__day-name"
              >
                ke
              </div>
              <div
                className="react-datepicker__day-name"
              >
                to
              </div>
              <div
                className="react-datepicker__day-name"
              >
                pe
              </div>
              <div
                className="react-datepicker__day-name"
              >
                la
              </div>
              <div
                className="react-datepicker__day-name"
              >
                su
              </div>
            </div>
          </div>
          <div
            aria-label="month-2019-12"
            className="react-datepicker__month"
            onMouseLeave={[Function]}
            role="listbox"
          >
            <div
              className="react-datepicker__week"
            >
              <div
                aria-disabled={false}
                aria-label="day-25"
                className="react-datepicker__day react-datepicker__day--025 react-datepicker__day--in-range react-datepicker__day--outside-month"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                25
              </div>
              <div
                aria-disabled={false}
                aria-label="day-26"
                className="react-datepicker__day react-datepicker__day--026 react-datepicker__day--in-range react-datepicker__day--outside-month"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                26
              </div>
              <div
                aria-disabled={false}
                aria-label="day-27"
                className="react-datepicker__day react-datepicker__day--027 react-datepicker__day--in-range react-datepicker__day--outside-month"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                27
              </div>
              <div
                aria-disabled={false}
                aria-label="day-28"
                className="react-datepicker__day react-datepicker__day--028 react-datepicker__day--in-range react-datepicker__day--outside-month"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                28
              </div>
              <div
                aria-disabled={false}
                aria-label="day-29"
                className="react-datepicker__day react-datepicker__day--029 react-datepicker__day--in-range react-datepicker__day--outside-month"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                29
              </div>
              <div
                aria-disabled={false}
                aria-label="day-30"
                className="react-datepicker__day react-datepicker__day--030 react-datepicker__day--in-range react-datepicker__day--weekend react-datepicker__day--outside-month"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                30
              </div>
              <div
                aria-disabled={false}
                aria-label="day-1"
                className="react-datepicker__day react-datepicker__day--001 react-datepicker__day--in-range react-datepicker__day--weekend"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                1
              </div>
            </div>
            <div
              className="react-datepicker__week"
            >
              <div
                aria-disabled={false}
                aria-label="day-2"
                className="react-datepicker__day react-datepicker__day--002 react-datepicker__day--in-range"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                2
              </div>
              <div
                aria-disabled={false}
                aria-label="day-3"
                className="react-datepicker__day react-datepicker__day--003 react-datepicker__day--in-range"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                3
              </div>
              <div
                aria-disabled={false}
                aria-label="day-4"
                className="react-datepicker__day react-datepicker__day--004 react-datepicker__day--in-range"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                4
              </div>
              <div
                aria-disabled={false}
                aria-label="day-5"
                className="react-datepicker__day react-datepicker__day--005 react-datepicker__day--in-range"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                5
              </div>
              <div
                aria-disabled={false}
                aria-label="day-6"
                className="react-datepicker__day react-datepicker__day--006 react-datepicker__day--in-range"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                6
              </div>
              <div
                aria-disabled={false}
                aria-label="day-7"
                className="react-datepicker__day react-datepicker__day--007 react-datepicker__day--in-range react-datepicker__day--weekend"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                7
              </div>
              <div
                aria-disabled={false}
                aria-label="day-8"
                className="react-datepicker__day react-datepicker__day--008 react-datepicker__day--in-range react-datepicker__day--weekend"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                8
              </div>
            </div>
            <div
              className="react-datepicker__week"
            >
              <div
                aria-disabled={false}
                aria-label="day-9"
                className="react-datepicker__day react-datepicker__day--009 react-datepicker__day--in-range"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                9
              </div>
              <div
                aria-disabled={false}
                aria-label="day-10"
                className="react-datepicker__day react-datepicker__day--010 react-datepicker__day--in-range"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                10
              </div>
              <div
                aria-disabled={false}
                aria-label="day-11"
                className="react-datepicker__day react-datepicker__day--011 react-datepicker__day--in-range"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                11
              </div>
              <div
                aria-disabled={false}
                aria-label="day-12"
                className="react-datepicker__day react-datepicker__day--012 react-datepicker__day--in-range"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                12
              </div>
              <div
                aria-disabled={false}
                aria-label="day-13"
                className="react-datepicker__day react-datepicker__day--013 react-datepicker__day--in-range"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                13
              </div>
              <div
                aria-disabled={false}
                aria-label="day-14"
                className="react-datepicker__day react-datepicker__day--014 react-datepicker__day--in-range react-datepicker__day--weekend"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                14
              </div>
              <div
                aria-disabled={false}
                aria-label="day-15"
                className="react-datepicker__day react-datepicker__day--015 react-datepicker__day--in-range react-datepicker__day--weekend"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                15
              </div>
            </div>
            <div
              className="react-datepicker__week"
            >
              <div
                aria-disabled={false}
                aria-label="day-16"
                className="react-datepicker__day react-datepicker__day--016 react-datepicker__day--in-range"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                16
              </div>
              <div
                aria-disabled={false}
                aria-label="day-17"
                className="react-datepicker__day react-datepicker__day--017 react-datepicker__day--in-range"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                17
              </div>
              <div
                aria-disabled={false}
                aria-label="day-18"
                className="react-datepicker__day react-datepicker__day--018 react-datepicker__day--in-range"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                18
              </div>
              <div
                aria-disabled={false}
                aria-label="day-19"
                className="react-datepicker__day react-datepicker__day--019 react-datepicker__day--in-range"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                19
              </div>
              <div
                aria-disabled={false}
                aria-label="day-20"
                className="react-datepicker__day react-datepicker__day--020 react-datepicker__day--selected react-datepicker__day--range-end react-datepicker__day--in-range"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                20
              </div>
              <div
                aria-disabled={false}
                aria-label="day-21"
                className="react-datepicker__day react-datepicker__day--021 react-datepicker__day--weekend"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                21
              </div>
              <div
                aria-disabled={false}
                aria-label="day-22"
                className="react-datepicker__day react-datepicker__day--022 react-datepicker__day--weekend"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                22
              </div>
            </div>
            <div
              className="react-datepicker__week"
            >
              <div
                aria-disabled={false}
                aria-label="day-23"
                className="react-datepicker__day react-datepicker__day--023"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                23
              </div>
              <div
                aria-disabled={false}
                aria-label="day-24"
                className="react-datepicker__day react-datepicker__day--024"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                24
              </div>
              <div
                aria-disabled={false}
                aria-label="day-25"
                className="react-datepicker__day react-datepicker__day--025"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                25
              </div>
              <div
                aria-disabled={false}
                aria-label="day-26"
                className="react-datepicker__day react-datepicker__day--026"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                26
              </div>
              <div
                aria-disabled={false}
                aria-label="day-27"
                className="react-datepicker__day react-datepicker__day--027"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                27
              </div>
              <div
                aria-disabled={false}
                aria-label="day-28"
                className="react-datepicker__day react-datepicker__day--028 react-datepicker__day--weekend"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                28
              </div>
              <div
                aria-disabled={false}
                aria-label="day-29"
                className="react-datepicker__day react-datepicker__day--029 react-datepicker__day--weekend"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                29
              </div>
            </div>
            <div
              className="react-datepicker__week"
            >
              <div
                aria-disabled={false}
                aria-label="day-30"
                className="react-datepicker__day react-datepicker__day--030"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                30
              </div>
              <div
                aria-disabled={false}
                aria-label="day-31"
                className="react-datepicker__day react-datepicker__day--031"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                31
              </div>
              <div
                aria-disabled={false}
                aria-label="day-1"
                className="react-datepicker__day react-datepicker__day--001 react-datepicker__day--outside-month"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                1
              </div>
              <div
                aria-disabled={false}
                aria-label="day-2"
                className="react-datepicker__day react-datepicker__day--002 react-datepicker__day--outside-month"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                2
              </div>
              <div
                aria-disabled={false}
                aria-label="day-3"
                className="react-datepicker__day react-datepicker__day--003 react-datepicker__day--outside-month"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                3
              </div>
              <div
                aria-disabled={false}
                aria-label="day-4"
                className="react-datepicker__day react-datepicker__day--004 react-datepicker__day--weekend react-datepicker__day--outside-month"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                4
              </div>
              <div
                aria-disabled={false}
                aria-label="day-5"
                className="react-datepicker__day react-datepicker__day--005 react-datepicker__day--weekend react-datepicker__day--outside-month"
                onClick={[Function]}
                onMouseEnter={[Function]}
                role="option"
              >
                5
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
